### PR TITLE
Parse reference disk manifest files and then mount reference disks on PAPI worker VMs if needed [BA-6326]

### PR DESCRIPTION
**Pull request by grsterin**

### CODE REVIEW

File: supportedBackends/google/pipelines/common/src/main/scala/cromwell/backend/google/pipelines/common/PipelinesApiReferenceFilesMapping.scala [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+      validReferenceFilesFromManifestMap map {
+        case (refFile, disk) => (refFile.path, disk)
+      }
+    }
+  }
+}
+
+object PipelinesApiReferenceFilesMapping {
+  case class ReferenceFile(path: String, crc32c: Long)
+  case class ManifestFile(imageIdentifier: String, diskSizeGb: Int, files: List[ReferenceFile])
```
Comment (cjllanwarne) [inline supportedBackends/google/pipelines/common/src/main/scala/cromwell/backend/google/pipelines/common/PipelinesApiReferenceFilesMapping.scala:139]: These should all be `final case class`es throughout

File: supportedBackends/google/pipelines/common/src/main/scala/cromwell/backend/google/pipelines/common/PipelinesApiReferenceFilesMapping.scala [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+    val validReferenceFilesFromManifestMapIo = filesWithValidatedCrc32csIo map {
+      filesWithValidatedCrc32c => {
+        allReferenceFilesFromManifestMap.filterKeys(key => filesWithValidatedCrc32c.getOrElse(key, false))
+      }
+    }
+
+    validReferenceFilesFromManifestMapIo map { validReferenceFilesFromManifestMap =>
+      val invalidReferenceFiles = allReferenceFilesFromManifestMap.keySet -- validReferenceFilesFromManifestMap.keySet
+      if (invalidReferenceFiles.nonEmpty) {
+        logger.warn(s"The following files listed in references manifest have checksum mismatch with actual files in GCS: ${invalidReferenceFiles.mkString(",")}")
```
Comment (cjllanwarne) [inline supportedBackends/google/pipelines/common/src/main/scala/cromwell/backend/google/pipelines/common/PipelinesApiReferenceFilesMapping.scala:128]: If I'm reading this right, it looks really nice - so if a reference is no longer valid, we get a warning about it, but it doesn't stop us from using the rest of the (still-valid) references on the disk?

Comment (grsterin) [reply supportedBackends/google/pipelines/common/src/main/scala/cromwell/backend/google/pipelines/common/PipelinesApiReferenceFilesMapping.scala:128]: Yes, this is correct

File: supportedBackends/google/pipelines/common/src/main/scala/cromwell/backend/google/pipelines/common/PipelinesApiReferenceFilesMapping.scala [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+    if (filesWithInvalidPaths.nonEmpty) {
+      IO.raiseError(new RuntimeException(s"Some of the paths in manifest file are not valid GCS paths: " +
+        s"${filesWithInvalidPaths.keySet.map(_.path).toList.mkString(", ")}")
+      )
+    } else {
+      IO {
+        val gcsBatch = gcsClient.batch()
+        val filesAndBlobResults = filesWithValidPaths.map {
+          case (referenceFile, ValidFullGcsPath(bucket, path)) =>
+            val blobGetResult = gcsBatch.get(BlobId.of(bucket, path.substring(1)), BlobGetOption.fields(BlobField.CRC32C))
```
[the remark this answers is not in the document]
Comment (grsterin) [reply supportedBackends/google/pipelines/common/src/main/scala/cromwell/backend/google/pipelines/common/PipelinesApiReferenceFilesMapping.scala:95]: I was successfully testing it with 1604 requests. So maybe Java client library does splitting transparently for consumer. Also client lib documentation doesn't mention any limits. I'll double check.

File: supportedBackends/google/pipelines/common/src/main/scala/cromwell/backend/google/pipelines/common/PipelinesApiReferenceFilesMapping.scala [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+        }
+    }
+    validReferenceFilesMapIO.unsafeRunSync()
+  }
+
+  def getReferenceDisksToMount(inputFilePaths: Set[String]): List[PipelinesApiReferenceFilesDisk] = {
+    validReferenceFilesMap.filterKeys(key => inputFilePaths.contains(s"gs://$key")).values.toList.distinct
+  }
+
+  private def createGcsClient(): Storage = StorageOptions
```
[the remark this answers is not in the document]
Comment (grsterin) [reply supportedBackends/google/pipelines/common/src/main/scala/cromwell/backend/google/pipelines/common/PipelinesApiReferenceFilesMapping.scala:50]: We don't, but I wanted those clients to be garbage collected once they are not needed anymore.

File: supportedBackends/google/pipelines/common/src/main/scala/cromwell/backend/google/pipelines/common/PipelinesApiReferenceFilesMapping.scala (hunk 3 above)
[the remark this answers is not in the document]
Comment (grsterin) [reply supportedBackends/google/pipelines/common/src/main/scala/cromwell/backend/google/pipelines/common/PipelinesApiReferenceFilesMapping.scala:95]: Confirmed, client library does all the dirty work of splitting requests into batches of 100 for you: https://github.com/googleapis/java-storage/blob/f8d6e158a06aec926fb7bc42f10483d56696a37e/google-cloud-storage/src/main/java/com/google/cloud/storage/spi/v1/HttpStorageRpc.java#L180

File: supportedBackends/google/pipelines/common/src/main/scala/cromwell/backend/google/pipelines/common/io/PipelinesApiAttachedDisk.scala [hunk 5]
```diff
@@ -81,7 +81,13 @@ object PipelinesApiWorkingDisk {
...
-  val name = PipelinesApiWorkingDisk.Name
+  val mountPoint: Path = PipelinesApiWorkingDisk.MountPoint
+  val name: String = PipelinesApiWorkingDisk.Name
   override def toString: String = s"$name $sizeGb ${diskType.diskTypeName}"
 }
+
+case class PipelinesApiReferenceFilesDisk(image: String, sizeGb: Int) extends PipelinesApiAttachedDisk {
+  val mountPoint: Path = DefaultPathBuilder.get(s"/mnt/${image.md5Sum}")
+  val name: String = s"d-${mountPoint.pathAsString.md5Sum}"
+  val diskType: DiskType = DiskType.SSD
```
[the remark this answers is not in the document]
Comment (grsterin) [reply supportedBackends/google/pipelines/common/src/main/scala/cromwell/backend/google/pipelines/common/io/PipelinesApiAttachedDisk.scala:92]: Added question to techtalk agenda

File: supportedBackends/google/pipelines/common/src/main/scala/cromwell/backend/google/pipelines/common/PipelinesApiReferenceFilesMapping.scala (hunk 1 above)
Comment (grsterin) [reply supportedBackends/google/pipelines/common/src/main/scala/cromwell/backend/google/pipelines/common/PipelinesApiReferenceFilesMapping.scala]: After replacing constructor initialization with case class and `generateReferenceFilesMapping` method in the companion object per suggestion above, I finally had to extract logic into a trait to make it testable, but now, when these 2 classes are inside a trait, it's impossible to make them final - it results in compilation error `Error:(36, 20) The outer reference in this type test cannot be checked at run time`. So we either leave them as is, or there should be another way to test the  logic.

File: supportedBackends/google/pipelines/common/src/main/scala/cromwell/backend/google/pipelines/common/PipelinesApiReferenceFilesMapping.scala [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+    val refDisk = PipelinesApiReferenceFilesDisk(manifestFile.imageIdentifier, manifestFile.diskSizeGb)
+    val allReferenceFilesFromManifestMap = manifestFile.files.map(refFile => (refFile, refDisk)).toMap
+    val referenceFilesWithValidPathsIo = getReferenceFileToValidatedGcsPathMap(allReferenceFilesFromManifestMap.keySet)
+    val filesWithValidatedCrc32csIo = bulkValidateCrc32cs(gcsClient, referenceFilesWithValidPathsIo)
+
+    val validReferenceFilesFromManifestMapIo = filesWithValidatedCrc32csIo map {
+      filesWithValidatedCrc32c => {
+        allReferenceFilesFromManifestMap.filterKeys(key => filesWithValidatedCrc32c.getOrElse(key, false))
+      }
+    }
```
Comment (mcovarr) [inline supportedBackends/google/pipelines/common/src/main/scala/cromwell/backend/google/pipelines/common/PipelinesApiReferenceFilesMapping.scala:130]: TOL could the valid / invalid groupings be generated more easily with `partition`?

File: supportedBackends/google/pipelines/common/src/test/scala/cromwell/backend/google/pipelines/common/PipelinesApiReferenceFilesMappingSpec.scala [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+            case ValidFullGcsPath("bucketname2", "manifest2") =>
+              ManifestFile(
+                imageIdentifier = disk2.image,
+                diskSizeGb = 100,
+                files = List(
+                  ReferenceFile(path = refFile3Disk2, crc32c = 12L),
+                  ReferenceFile(path = refFile4Disk2MismatchingChecksum, crc32c = 12345L),
+                )
+              )
+            case _ => throw new RuntimeException("Exception in mock object. This should not happen")
```
Comment (mcovarr) [inline supportedBackends/google/pipelines/common/src/test/scala/cromwell/backend/google/pipelines/common/PipelinesApiReferenceFilesMappingSpec.scala:64]: Putting whatever this was into the message might save some time debugging. 🙂

File: supportedBackends/google/pipelines/common/src/test/scala/cromwell/backend/google/pipelines/common/PipelinesApiConfigurationAttributesSpec.scala [hunk 8]
```diff
@@ -401,13 +403,4 @@ class PipelinesApiConfigurationAttributesSpec extends FlatSpec with Matchers {
...
 
-  it should "parse correct reference-disk-localization-manifest-files config" in {
-    val manifest1Path = "gs://bucket/manifest1.json"
-    val manifest2Path = "gs://bucket/manifest2.json"
-    val manifestConfigStr = s"""reference-disk-localization-manifest-files = ["$manifest1Path", "$manifest2Path"]""".stripMargin
-    val backendConfig = ConfigFactory.parseString(configString(manifestConfigStr))
-
-    val pipelinesApiAttributes = PipelinesApiConfigurationAttributes(googleConfig, backendConfig, "papi")
-    pipelinesApiAttributes.referenceDiskLocalizationManifestFiles should be(Option(List(manifest1Path, manifest2Path)))
-  }
```
Comment (mcovarr) [inline supportedBackends/google/pipelines/common/src/test/scala/cromwell/backend/google/pipelines/common/PipelinesApiConfigurationAttributesSpec.scala:412]: ooc why was this removed?

File: supportedBackends/google/pipelines/common/src/main/scala/cromwell/backend/google/pipelines/common/PipelinesApiReferenceFilesMapping.scala (hunk 3 above)
[the remark this answers is not in the document]
Comment (grsterin) [reply supportedBackends/google/pipelines/common/src/main/scala/cromwell/backend/google/pipelines/common/PipelinesApiReferenceFilesMapping.scala]: This also makes me wonder if we have some redundant queuing logic in the `IoActor`: https://github.com/broadinstitute/cromwell/blob/ff139e809df248b5aa8cd7cc8c6de2b153b0c807/engine/src/main/scala/cromwell/engine/io/IoActor.scala#L60

File: supportedBackends/google/pipelines/common/src/test/scala/cromwell/backend/google/pipelines/common/PipelinesApiConfigurationAttributesSpec.scala (hunk 8 above)
Comment (grsterin) [reply supportedBackends/google/pipelines/common/src/test/scala/cromwell/backend/google/pipelines/common/PipelinesApiConfigurationAttributesSpec.scala:412]: added it back, and added some more

File: supportedBackends/google/pipelines/common/src/main/scala/cromwell/backend/google/pipelines/common/PipelinesApiReferenceFilesMapping.scala [hunk 9]
```diff
@@ -0,0 +1,137 @@
...
+      jsonStringIo.flatMap(jsonStr => IO.fromEither(decode[ManifestFile](jsonStr)))
+    }
+  }
+
+  private def getReferenceFileToValidatedGcsPathMap(referenceFiles: Set[ReferenceFile]): IO[Map[ReferenceFile, ValidFullGcsPath]] = {
+    val filesAndValidatedPaths = referenceFiles.map {
+      referenceFile => (referenceFile, GcsPathBuilder.validateGcsPath(s"gs://${referenceFile.path}"))
+    }.toMap
+
+    val (filesWithValidPaths, filesWithInvalidPaths) = filesAndValidatedPaths.partition(_._2.isInstanceOf[ValidFullGcsPath])
```
Comment (cjllanwarne) [inline supportedBackends/google/pipelines/common/src/main/scala/cromwell/backend/google/pipelines/common/PipelinesApiReferenceFilesMapping.scala:79]: I actually almost preferred the dual `.collect`s version you had before because it guaranteed the types for you and avoided the awkward `_.asInstanceOf`s. 

I wonder whether there's an equivalent that does a single pass over a list but divides into two well-typed outputs...

UPDATE: based on quick stackoverflow searching, this seems like the closest thing: https://stackoverflow.com/a/24216337

File: supportedBackends/google/pipelines/common/src/main/scala/cromwell/backend/google/pipelines/common/PipelinesApiReferenceFilesMapping.scala [hunk 10]
```diff
@@ -0,0 +1,137 @@
...
+
+    if (filesWithInvalidPaths.nonEmpty) {
+      IO.raiseError(new InvalidGcsPathsInManifestFileException(filesWithInvalidPaths.keySet.map(_.path).toList))
+    } else {
+      IO.pure(filesWithValidPaths.mapValues(_.asInstanceOf[ValidFullGcsPath]))
```
Comment (cjllanwarne) [inline supportedBackends/google/pipelines/common/src/main/scala/cromwell/backend/google/pipelines/common/PipelinesApiReferenceFilesMapping.scala:84]: IO.pure followed by something which might throw an exception makes me nervous - but I think doing the well-typed-lists fix above should remove that nervousness.

File: supportedBackends/google/pipelines/common/src/main/scala/cromwell/backend/google/pipelines/common/PipelinesApiReferenceFilesMapping.scala [hunk 11]
```diff
@@ -0,0 +1,137 @@
...
+    }
+  }
+
+  protected def bulkValidateCrc32cs(gcsClient: Storage,
+                          filesWithValidPathsIo: IO[Map[ReferenceFile, ValidFullGcsPath]]): IO[Map[ReferenceFile, Boolean]] = {
```
Comment (cjllanwarne) [inline supportedBackends/google/pipelines/common/src/main/scala/cromwell/backend/google/pipelines/common/PipelinesApiReferenceFilesMapping.scala:89]: In general I find it's often neater (and easier to read each method's logic) to pass arguments in as full values (eg `filesWithValidPathsIo: Map[ReferenceFile, ValidFullGcsPath]`) and do the `IO` juggling with `flatMap`s at the point that you're calling the method

File: supportedBackends/google/pipelines/common/src/main/scala/cromwell/backend/google/pipelines/common/PipelinesApiReferenceFilesMapping.scala [hunk 12]
```diff
@@ -0,0 +1,137 @@
...
+    val validReferenceFilesFromManifestMapIo = filesWithValidatedCrc32csIo map {
+      filesWithValidatedCrc32c => {
+        allReferenceFilesFromManifestMap.filterKeys(key => filesWithValidatedCrc32c.getOrElse(key, false))
+      }
+    }
+
+    validReferenceFilesFromManifestMapIo map { validReferenceFilesFromManifestMap =>
+      val invalidReferenceFiles = allReferenceFilesFromManifestMap.keySet -- validReferenceFilesFromManifestMap.keySet
+      if (invalidReferenceFiles.nonEmpty) {
+        logger.warn(s"The following files listed in references manifest have checksum mismatch with actual files in GCS: ${invalidReferenceFiles.mkString(",")}")
```
Comment (cjllanwarne) [inline supportedBackends/google/pipelines/common/src/main/scala/cromwell/backend/google/pipelines/common/PipelinesApiReferenceFilesMapping.scala:130]: Could we make a ticket in the same epic to add an alert for this in production?

File: supportedBackends/google/pipelines/common/src/main/scala/cromwell/backend/google/pipelines/common/PipelinesApiReferenceFilesMapping.scala (hunk 1 above)
Comment (cjllanwarne) [reply supportedBackends/google/pipelines/common/src/main/scala/cromwell/backend/google/pipelines/common/PipelinesApiReferenceFilesMapping.scala]: Did you try putting these into a static `object` instead of a trait? I'm not sure what happens when you try to access trait-defined classes from different subtypes of that trait... it seems like maybe you just cannot!

Comment (grsterin) [reply supportedBackends/google/pipelines/common/src/main/scala/cromwell/backend/google/pipelines/common/PipelinesApiReferenceFilesMapping.scala]: Currently they are inside the trait, but it's impossible to make them private.

File: supportedBackends/google/pipelines/common/src/main/scala/cromwell/backend/google/pipelines/common/PipelinesApiReferenceFilesMapping.scala (hunk 9 above)
Comment (grsterin) [reply supportedBackends/google/pipelines/common/src/main/scala/cromwell/backend/google/pipelines/common/PipelinesApiReferenceFilesMapping.scala:79]: I don't like the `foldRight` approach since it's not immediately obvious what's happening there, so every time while reading the code I'd have to stop and think.

File: supportedBackends/google/pipelines/common/src/main/scala/cromwell/backend/google/pipelines/common/PipelinesApiReferenceFilesMapping.scala (hunk 12 above)
Comment (grsterin) [reply supportedBackends/google/pipelines/common/src/main/scala/cromwell/backend/google/pipelines/common/PipelinesApiReferenceFilesMapping.scala:130]: https://broadworkbench.atlassian.net/browse/BA-6574

File: supportedBackends/google/pipelines/common/src/main/scala/cromwell/backend/google/pipelines/common/PipelinesApiReferenceFilesMapping.scala [hunk 13]
```diff
@@ -71,57 +71,58 @@ protected trait PipelinesApiReferenceFilesMappingOperations {
...
+    val filesWithValidPaths = filesAndValidatedPaths.collect {
+      case (referenceFile, validPath: ValidFullGcsPath) => (referenceFile, validPath)
+    }
+    val filesWithInvalidPaths = filesAndValidatedPaths.collect {
+      case (referenceFile, invalidPath: InvalidFullGcsPath) => (referenceFile, invalidPath)
+    }
 
     if (filesWithInvalidPaths.nonEmpty) {
-      IO.raiseError(new InvalidGcsPathsInManifestFileException(filesWithInvalidPaths.keySet.map(_.path).toList))
+      throw new InvalidGcsPathsInManifestFileException(filesWithInvalidPaths.keySet.map(_.path).toList)
```
Comment (cjllanwarne) [inline supportedBackends/google/pipelines/common/src/main/scala/cromwell/backend/google/pipelines/common/PipelinesApiReferenceFilesMapping.scala]: I'm not sure whether this change was caused by me saying I would prefer function inputs to be simple types instead of `IO`s. If so, sorry, I didn't mean to imply that throwing exceptions is better than wrapping these unsafe operations in `IO`s. I think we should almost never throw exceptions directly and almost always prefer `IO`s, especially in `trait` functions like these.

All I meant was that we should do the `flatMap`s above functions rather than within them, so that each conversion is written as a function of a simple type, and we can compose them together if we need to. Eg instead of 
```
def functionA(t1: Type1): IO[Type2] = ...
def functionB(t2: IO[Type2]): IO[Type3] = ...

def functionC(t1: Type1): IO[Type3] = functionB(functionA(t1))
```

I would prefer:
```
def functionA(t1: Type1): IO[Type2] = ...
def functionB(t2: Type2): IO[Type3] = ...

def functionC(t1: Type1): IO[Type3] = for {
  t2 <- functionA(t1)
  t3 <- functionB(t2)
} yield t3
```
